[RTC-7] WebRTC congestion control (#1)
jellyfish-dev/book: src/webrtc/gcc.md

<!-- 1. Congestion Control
2. TWCC vs GCC
3. GCC
4. GCC & Simulcast
5. Connection Probing
6. Strategies for probing the connection -->

## Congestion Control

The network has its capacity. When a sender sends more data than network devices are able to process,
packets will be delayed and finally dropped. Such situation is called the network congestion.

Congestion control goal is to prevent from congesting the network by controling the sender rate.
Every time, the sender notices that the packet transmission time increases or packets are getting
lost it should lower its rate.

<!-- In the case of multimedia, lowering sender rate is challenging as the audio and video 
have specified bitrate and sometimes even lowering the video quality might not  -->

## TWCC vs GCC

TWCC (Transport-wide Congestion Control) defines new RTP header extension and RTCP message used to 
feedback the arrival times of RTP packets to the sender.
Each RTP packet is assigned transport-wide (unique across all RTP streams sent at the same socket) 
sequence number by the sender before transmission.
The receiver records arrival timestamps of packets and send them back in TWCC Feedback message.

> Then, the sender is able to determine network delay and packet loss, and react to them by lowering its rate.

Reviewer: I think the first part of the sentence doesn't make it clear that we're talking about significant network delay and significant packet loss.
I also have a problem with `its rate`. You didn't introduce the sender rate yet. Let's do `amount of data it sends`, shall we?
Maybe we should rewrite it as...
```suggestion
Then the sender is able to determine network delay and packet loss and react to them by adjusting the amount of data it sends, increasing bitrate if network conditions are good or decreasing it if they aren't.
```

Also, a pro tip, use Grammarly or something. In English, a comma is used rather sparsely and it will highlight the misuse of commas, as well as other useful things.

Author: >  In English, a comma is used rather sparsely and it will highlight the misuse of commas, as well as other useful things.

I will, this PR wasn't meant for CR yet :D

Author: > I think the first part of the sentence doesn't make it clear that we're talking about significant network delay and significant packet loss.

That's only an introduction/general description. The details (e.g. packet loss threshold triggering the reaction) can be provided later

> I also have a problem with its rate. You didn't introduce the sender rate yet. Let's do amount of data it sends, shall we?

You mean that it's unclear what `its` refers to? The `sender rate` was used in the `Congestion Control` section for the first time

How this reaction looks like, i.e. how much sender should lower its rate, when it should increase and when 
it should hold its rate, etc. is out-of-scope of TWCC and depends on specific congestion control algorithm.

> In other words, TWCC provides only means to implement actual congestion control algorithm.

Reviewer: ```suggestion
In other words, TWCC only provides the means necessary to implement an actual congestion control algorithm, eg. GCC.
```

> GCC (Google Congestion Control) is actual congestion control algorithm developed by Google.
> It uses TWCC Feedback messages to determine when and how much to increase and decrease sender rate, 
> or when the sender rate should remain at the current level.

Reviewer: ```suggestion
GCC (Google Congestion Control) is an actual congestion control algorithm developed by Google.
It uses TWCC Feedback messages to determine when and how much to increase and decrease the sender rate, 
or when the sender rate should remain at the current level.
```

## GCC

GCC goal is to estimate the rate that sender can send its data with.
The whole system depends on over-using the channel i.e. to get higher estimation we have to increase our rate.

> From the RFC draft:

Reviewer: Please add an RFC number and a section number. Preferably as a link.

> the system depends on over-using the channel to verify the current available bandwidth estimate, we must make
> sure that our estimate does not diverge from the rate at which the sender is actually sending.  
> Thus, if the sender is unable to produce a bit stream with the bitrate the congestion controller is asking for, 
> the available bandwidth estimate should stay within a given bound.

The bandwidth estimation provided by congestion controller is in fact a request to the sender for producing 
the stream with specific bitrate.
If the network behavior remains unchanged after increasing sender rate the estimation is increased.
If the network starts being congested (transmission delay increases, packets get lost), the estimation
is decreased.
This way we are slowly increasing the sender rate up to the point where the network can't accept more data.

There are two increase modes: multiplicative and additive.

> In the multiplicative mode, the estimation is increase max by 8% per second.

Reviewer: ```suggestion
In the multiplicative mode, the estimation is increased by no more than 8% per second.
```

> In the additvie mode, the estimation is increased with at most half a packet per `response_time_interval` (RTT + 100ms).

Reviewer: I'm not sure how to interpret the second part of the sentence. But I think I would write similarly to the previous one.
```suggestion
In the additive mode, the estimation can be increased by no more than half a packet per `response_time_interval` (RTT + 100ms).
```

Author: I took it from the RFC :roll_eyes:

Reviewer: There might be room for improvement over the RFC :roll_eyes:

In other words, additive mode is much slower and is used when the estimation is close to convergence.

> When estimation is decreased, it is set to 85% of the latest receiver rate.

Reviewer: Is it accurate? I recall that we were talking about this behavior in the context of overusing the network. Isn't there a way for the additive mode to decrease the estimation a little more gently than that?

Author: Take a look at 

```
   When an over-use is detected the system transitions to the decrease
   state, where the delay-based available bandwidth estimate is
   decreased to a factor times the currently incoming bitrate.

     A_hat(i) = beta * R_hat(i)

   beta is typically chosen to be in the interval [0.8, 0.95], 0.85 is
   the RECOMMENDED value.
   ```

The questions we need to answer are:
1. what data use for probing
1. how much to probe
1. how often to probe
1. how long to probe

### 1. What data use for probing

The naive implementation would be to switch to the higher layer every `x` seconds and observe the network.
This however, may lead to periodic video rebuffering and lower Quality of Experience (QoE).

Another solution is to generate traffic by sending RTP padding packets.
In this case, server has full control over the amout of data being sent and therefore, it can probe with 
different aggresivness.

The problem with padding packets is that they contain only zeros (except the last byte) so we end up sending
a lot of garbage data, useless for the client. 
Fortunatelly, there is a solution which is called RTX.
RTX is a shortcut for retransmissions and can be used for probing instead of padding packets.
How does it work?
Instead of sending padding packets we simply retranssmit some packets via dedicated RTP stream.
This way we have a full control over probing rate and we send data that is useful for the client.
That's also the way libwebrtc chooses when RTX extension is enabled.

#### Probing with padding packets

> In our webrtc implementation, we started with padding packets as they are easier for implementing and give

Reviewer: ```suggestion
In our webrtc implementation, we started with padding packets as they are easier to implement and give
```

us pretty good results.
Here are some tips important when using padding packets:
* the `P` bit in the RTP header has to be set to `1`.
It indicates that RTP packet contains padding at the end of a payload.  

> * maximal size of padding only payload is 255.

Reviewer: ```suggestion
* maximum size of padding only payload is 255.
```
https://grammarhow.com/maximal-vs-maximum/

The limitation comes from the last byte. When the `P` flag is set, the last byte in payload denotes number of 

> zeros that should be stripped out.

Reviewer: It just occurred to me. Wouldn't we want to introduce a concept of padding packets separately from probing?

Author: Good point. I will introduce this term befor going into tips

* padding packets can only be sent at video frame boundary.
In other case the decoder in a web browser might get out-of-sync and you will see a lot of freezes.
How can you know when a video frames ends?
Just check the `M` bit in the RTP header.
For most codecs, `M` bit set to `1` means the end of a video frame.
* you have to take care of rewriting sequence numbers
* as a timestamp for padding packet you can use the timestamp of the last RTP packet with `M` bit set
* sending padding packets will cause your SRTP window to move forward. 
This might be problematic for out-of-order packets making it impossible to encrypt them as their sequence number
will not fit into the encryption window.
To solve this, you can either carefully increase the encryption window (too big encryption window might
introduce security vulnerabilities) or just drop the packet.

### 2. How much to probe

The estimation can increase maximum by 8% per second.

> This means that there is no sense in probing much more than your estimation can increase.

Reviewer: Nice point. Shouldn't we then rewrite some of our probing targets to target, let's say, 110% of GCC estimation?

Author: Sounds good, I am curious the results. In particular whether there will be any performance gains :thinking:
